doc/y_cable/design_doc.md

```diff
@@ -0,0 +1,247 @@
+
+  - Currently the assumption is Y-cable API will be written in the Python language
+  - If a vendor has an existing library in a different language, they will need to either find a way to wrap/bind it in Python to align with the description below or (preferrably) provide a pure Python implementation
+  - Since sonic_y_cable (sonic-platform-common/sonic_y_cable) is already built as a package today (only for a single Y cable vendor with sonic-buildimage) vendors can also place their implementation in this directory itself.
+
+#### Proposed Solution
+
+- Vendors can place their implementations in this format
+
+```
+	sonic_platform_common/sonic_y_cable/<vendor>/y_cable.py
+```
+- For example
+
+```
+	sonic_platform_common/sonic_y_cable/credo/y_cable.py
+```
+#### Rationale
+- The requirement here would be that the vendors must create their modules such that It can easily be accessed/imported and the modules also adhere to a uniform convention
+
+  - Vendor name should be derivable from the vendor name field inside the register spec of transceiver type.
+  - This is essential so that host/PMon container can determine the vendor name from eeprom and import the module of the transceiver appropriately
+
+
+### Port to module mapping
+- Another thing that is required is once we do have a vendor to module mapping, we need to map appropriate port to a module as well
+
+#### Background
+
+  - Basically, the key idea is once we have a port identified as being of a certain vendor and it has been identified to load a certain module which we can gather from the mapping described above we still need to call the correct module on each port each time we call the API on the port
+  - want to maintain this mapping in memory since xcvrd does not want to read/parse this y_cable_vendor_mapping.csv file again and again each time we call the Y-cable API
+  - Also note that the module loaded might change during xcvrd running lifetime since cable can be changed from one vendor to another. So we need to take this into consideration as well
+
+#### Proposed Solution
+
+  - Each module of the Y cable vendor can be a class (of each transceiver type) and all we need to do is instantiate the objects of these classes as class instances and these objects will provide the interface of calling the API's for the appropriate vendor Y cable.
+  - This instantiation will be done inside xcvrd, when xcvrd starts
+  - These objects basically emulate Y cable instances and whatever action/interaction needs to be done with the Ycable the methods of these objects would provide that
+  - each vendor in their implementation can inherit from a base class where there will be definitions for all the supported capabilities of the Y-cable.
+  - If the vendor does not support or implement the utility, it will just raise a not implemented error from the base class itself
```

Comment:
It is suggested by the Python community that if a subclass does not support a method, it should set the method equal to None. This differentiates it from a method they forgot to implement. Then, the calling code should first check if the method is None before attempting to call it.

Thus, vendors' concrete implementations should _never_ raise `NotImplementedError`

Reply:
left a note mentioning this, fixed